src/helperFunctions/process.py

```diff
@@ -21,8 +23,35 @@ def complete_shutdown(message: Optional[str] = None) -> None:
     if message is not None:
         logging.warning(message)
     logging.critical('SHUTTING DOWN SYSTEM')
-    process_group_id = os.getpgid(os.getpid())
-    os.killpg(process_group_id, SIGKILL)
+    _stop_remaining_fact_processes()
+
+
+def _stop_remaining_fact_processes():
+    """Find subprocesses of this process group and stop them."""
+    pgid = os.getpgrp()
+    futures = []
+    with ThreadPoolExecutor() as pool:
+        for proc in psutil.process_iter():
+            try:
+                if os.getpgid(proc.pid) == pgid and proc.pid != pgid:
+                    futures.append(pool.submit(_stop_process_by_pid, proc.pid))
+            except ProcessLookupError:
+                pass
+        for future in futures:
+            future.result()  # call result to make sure all threads are finished and there are no exceptions
+
+
+def _stop_process_by_pid(pid: int):
+    try:
+        proc = psutil.Process(pid)
+        try:
+            proc.wait(5)
+        except psutil.TimeoutExpired as err:
+            logging.warning(f'Timeout while waiting for process to shut down: {err} -> kill')
```

Comment:
Maybe add some more information to this log entry.
I'm not sure what exactly would be helpful, but it would be great to somehow be able to determine what process exactly was killed.

Reply:
We could print the pid but I don't think that this would really be useful after we killed it :sweat_smile: 
Sadly there seems to be no easy way to find out what the process was actually doing when we try to stop it